testsuite/oidc/test_client.py

```diff
@@ -0,0 +1,187 @@
+                refresh_token_func,
+                new_token_data.get("refresh_token", ""),
+            )
+
+        return refresh_token_func
+
+    @classmethod
+    def create_confidential_client(cls, keycloak, hostname: str, client_id: str = "my-confidential-client"):
+        """Factory method for confidential client"""
+        config = ClientConfig(
+            client_id=client_id,
+            client_type="confidential",
+            public_client=False,
+            redirect_uris=[f"http://{hostname}/*"],
+            web_origins=[f"http://{hostname}"],
+            root_url=f"http://{hostname}",
+            service_accounts_enabled=True,
+        )
+
+        keycloak_client = keycloak.realm.create_client(**config.to_keycloak_payload())
+
+        oidc_client = KeycloakOpenID(
+            server_url=keycloak.server_url,
+            client_id=keycloak_client.auth_id,
+            realm_name=keycloak.realm_name,
+            client_secret_key=keycloak_client.secret,
+        )
+
+        return cls(oidc_client, client_id, keycloak_client.secret), keycloak_client
+
+    @classmethod
+    def create_public_client(cls, keycloak, hostname: str, client_id: str = "my-public-client"):
+        """Factory method for public client"""
+        config = ClientConfig(
+            client_id=client_id,
+            client_type="public",
+            public_client=True,
+            redirect_uris=[f"http://{hostname}/*"],
+            web_origins=[f"http://{hostname}"],
+            root_url=f"http://{hostname}",
+            service_accounts_enabled=False,
+        )
+
+        keycloak_client = keycloak.realm.create_client(**config.to_keycloak_payload())
+
+        oidc_client = KeycloakOpenID(
+            server_url=keycloak.server_url,
+            client_id=keycloak_client.auth_id,
+            realm_name=keycloak.realm_name,
+        )
+
+        return cls(oidc_client, client_id, None), keycloak_client
+
+    @classmethod
+    def create_service_client(cls, keycloak, hostname: str, client_id: str = "my-service-client"):
+        """Factory method for service account client"""
+        config = ClientConfig(
+            client_id=client_id,
+            client_type="service",
+            public_client=False,
+            standard_flow_enabled=False,
+            service_accounts_enabled=True,
+            redirect_uris=[f"http://{hostname}/*"],
+            web_origins=[f"http://{hostname}"],
+            root_url=f"http://{hostname}",
+            direct_access_grants_enabled=False,
+        )
+
+        keycloak_client = keycloak.realm.create_client(**config.to_keycloak_payload())
+
+        oidc_client = KeycloakOpenID(
+            server_url=keycloak.server_url,
+            client_id=keycloak_client.auth_id,
+            realm_name=keycloak.realm_name,
+            client_secret_key=keycloak_client.secret,
+        )
+
+        return cls(oidc_client, client_id, keycloak_client.secret), keycloak_client
```

Comment:
I see some nice patterns here. Maybe you can create a class for each of the clients with it's own initialization method

Reply:
I think you have more duplication in methods here than if you had one base class with a single client function that will be reused by client classes that inherit the base one. 

I think it is a better structure, don't need to do this though